test/cipher.go

```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+
+		bc = newCipher(nkeys[i])
+		decrypted := make([]byte, len(text))
+		bc.Message(decrypted, ncrypts[i], ncrypts[i])
+		if !bytes.Equal(text, decrypted) {
+			t.Log("Encryption / Decryption failed", i)
+			t.FailNow()
+		}
+	}
+
+	for i := range ncrypts {
+		for j := i + 1; j < len(ncrypts); j++ {
+			if bytes.Equal(ncrypts[i], ncrypts[j]) {
+				t.Log("Different keys result in same encryption")
+				t.FailNow()
+			}
+
+			res := BitDiff(ncrypts[i], ncrypts[j])
+			if res < bitdiff {
+				t.Log("Encryptions not sufficiently different:", res)
+				t.FailNow()
+			}
+		}
+	}
+}
+
+// Tests a Cipher:
+// 1) Encryption / decryption work
+// 2) Encryption / decryption with different key don't work
+// 3) Changing a bit in the ciphertext or mac results in failed mac check
+// 4) Different keys produce sufficiently random output
+func AuthenticateAndEncrypt(t *testing.T,
+	newCipher func([]byte, ...interface{}) abstract.Cipher,
+	n int, bitdiff float64, text []byte) {
+	cryptsize := len(text)
+	decrypted := make([]byte, len(text))
+
+	bc := newCipher(nil)
+	keysize := bc.KeySize()
+	hashsize := bc.HashSize()
+	mac := make([]byte, hashsize)
+
+	ncrypts := make([][]byte, n)
+	nkeys := make([][]byte, n)
+	nmacs := make([][]byte, n)
+
+	// Encrypt / decrypt / mac test
+	for i := range nkeys {
+		nkeys[i] = make([]byte, keysize)
+		rand.Read(nkeys[i])
+		bc = newCipher(nkeys[i])
+		ncrypts[i] = make([]byte, cryptsize)
+		bc.Message(ncrypts[i], text, ncrypts[i])
+		nmacs[i] = make([]byte, hashsize)
+		bc.Message(nmacs[i], nil, nil)
+
+		bc = newCipher(nkeys[i])
+		bc.Message(decrypted, ncrypts[i], ncrypts[i])
+		if !bytes.Equal(text, decrypted) {
+			t.Log("Encryption / Decryption failed", i)
+			t.FailNow()
+		}
+
+		bc.Message(mac, nmacs[i], nil)
+		if subtle.ConstantTimeAllEq(mac, 0) != 1 {
+			t.Log("MAC Check failed")
+			t.FailNow()
+		}
+	}
+
+	// Different keys test
+	for i := range ncrypts {
+		for j := range ncrypts {
+			if i == j {
+				continue
+			}
+			bc = newCipher(nkeys[i])
+			bc.Message(decrypted, ncrypts[j], ncrypts[j])
+			bc.Message(mac, nmacs[j], nil)
+			if subtle.ConstantTimeAllEq(mac, 0) == 1 {
+				t.Log("MAC Check passed")
+				t.FailNow()
+			}
+		}
+	}
+
+	// Not enough randomness in 1 byte to pass this consistently
+	if len(ncrypts[0]) < 8 {
+		return
+	}
+
+	// Bit difference test
+	for i := range ncrypts {
+		for j := i + 1; j < len(ncrypts); j++ {
+			res := BitDiff(ncrypts[i], ncrypts[j])
+			if res < bitdiff {
+				t.Log("Encryptions not sufficiently different", res)
+				t.FailNow()
+			}
+		}
+	}
+
+	deltacopy := make([]byte, cryptsize)
+
+	// Bit flipping test
+	for i := range ncrypts {
+		copy(ncrypts[i], deltacopy)
+
+		deltacopy[0] ^= 255
+		bc = newCipher(nkeys[i])
+		bc.Message(decrypted, deltacopy, deltacopy)
+		bc.Message(mac, nmacs[i], nil)
+		if subtle.ConstantTimeAllEq(mac, 0) == 1 {
+			t.Log("MAC Check passed")
+			t.FailNow()
+		}
+		deltacopy[0] = ncrypts[i][0]
+
+		deltacopy[len(deltacopy)/2-1] ^= 255
```

Comment:
Looks like each of these blocks of code is mostly duplicate but testing a slightly different case; factor out the common code into a separate function.  Doesn't need to be a separate top-level function if an internal higher-order function def would be useful, i.e.,

```
deltatest := func (...) { ...}
```

very handy because you can also refer to internal variables in the outer test function from the inner higher-order function.